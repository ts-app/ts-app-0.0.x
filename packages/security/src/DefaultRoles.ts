export enum DefaultRoles {
  // support for self role...
  Administrator = 'Administrator',
  User = 'User'
}
