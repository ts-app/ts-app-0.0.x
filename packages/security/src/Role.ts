/**
 * This type specifies a name and unique ID for a Role.
 */
export interface Role {
  id: string
  name: string
}
