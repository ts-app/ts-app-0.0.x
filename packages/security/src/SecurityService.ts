import { Profile, User } from '@ts-app/common'
import { FindWithCursor } from '@ts-app/mongo'

export interface SecurityService {
  signUp (input: { email: string, password: string }): Promise<{ error?: string, user?: User }>

  loginWithEmailPassword (input: { email: string, password: string }): Promise<{ error?: string, userId?: string, accessToken?: string, refreshToken?: string }>

  user (id: string): Promise<{ error?: string, user?: User }>

  users (input: { q?: string, limit?: number, cursor?: string }): Promise<FindWithCursor<User>>

  removeUser (id: String): Promise<number>

  seedUsers (input?: { force?: boolean, userCount?: number }): Promise<{ error?: string }>

  updateProfile (input: { id: string, profile: object }): Promise<{ error?: string }>

  reset (): Promise<{ error?: string }>

  // --- TODO: not impl

  loginWithUserId (id: string): Promise<{ error?: string, user?: User }>

  setProfile<T extends Profile> (input: { id: string, profile: T }): Promise<void>

  getProfile<T extends Profile> (input: { id: string }): Promise<T>

}
