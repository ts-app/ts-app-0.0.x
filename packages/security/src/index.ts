export * from './DefaultRoles'
export * from './MongoRoleService'
export * from './MongoSecurityService'
export * from './Role'
export * from './RoleService'
export * from './securitySchemaDefinition'
export * from './SecurityService'
export * from './validators'
