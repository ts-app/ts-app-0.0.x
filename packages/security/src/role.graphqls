type Query {
    userIsInRoles(userId: String!, roles: [String]!, group: String): Boolean
    getGroupsForUser(userId: String, role: String): GetGroupsForUserPayload
    findRolesWithCursor(q: String, limit: Int, cursor: String): FindRolesWithCursorPayload
    userRoles(userId: String!, group: String): GetRolesForUserPayload
    getUsersInRoles(roles: [String]!, group: String, limit: Int, cursor: String): GetUsersInRolesPayload
}

type Mutation {
    addUsersToRoles(userIds: [String]!, roles: [String], group: String): NoPayload
    createRole(name: String!): CreateRolePayload
    removeRole(name: String!): NoPayload
    removeUsersFromRoles(userIds: [String]!, roles: [String]!, group: String): NoPayload
    removeUsersFromAllRoles(userIds: [String]!): NoPayload
}

type Role {
    id: String
    name: String
}

type CreateRolePayload {
    error: String
    id: String
}

type GetGroupsForUserPayload {
    error: String
    groups: [String]
}

type FindRolesWithCursorPayload {
    error: String
    cursor: String
    docs: [Role]
}

type GetRolesForUserPayload {
    error: String
    roles: [String]
}

type GetUsersInRolesPayload {
    error: String
    cursor: String
    docs: [User]
}
