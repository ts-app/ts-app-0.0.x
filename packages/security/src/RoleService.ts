import { Role } from './Role'
import { FindWithCursor } from '@ts-app/mongo'
import { User } from '@ts-app/common'

export const RoleServiceConstants = {
  GLOBAL: '_global_'
}

/**
 * Service to manage user access via roles and groups.
 *
 */
export interface RoleService {
  /**
   * Add users to specified roles, optionally specifying a group. If group name is not specified,
   * roles are assigned to RoleServiceConstants.GLOBAL group.
   *
   * @param {{userIds: (string | string[]); roles: (string | string[]); group?: string}} input
   * @return {Promise<{error?: string}>}
   */
  addUsersToRoles (input: {
    userIds: string | string[], roles: string | string[],
    group?: string
  }): Promise<{ error?: string }>

  /**
   * Create specified role name.
   *
   * @param {{name: string}} input
   * @return {Promise<{error?: string; id: string}>}
   */
  createRole (input: { name: string }): Promise<{ error?: string, id: string }>

  /**
   * Delete specified role name.
   *
   * If role is in use, service output error with appropriate message will be returned.
   *
   * @param {{name: string}} input
   * @return {Promise<{error?: string}>}
   */
  removeRole (input: { name: string }): Promise<{ error?: string }>

  /**
   * Get user's groups where user has role(s) assigned. RoleServiceConstants.GLOBAL will be omitted.
   *
   * If role is specified, restrict groups returned to those with specified role assigned.
   *
   * @param {{userId: string; role?: string}} input
   * @return {Promise<{error?: string; groups?: string[]}>}
   */
  getGroupsForUser (input: { userId: string, role?: string }): Promise<{ error?: string, groups?: string[] }>

  findRolesWithCursor (input?: {
    q?: string, limit?: number,
    cursor?: string
  }): Promise<FindWithCursor<Role>>

  /**
   * Get roles assigned to user, optionally restricting specifying a group. Roles assigned to
   * RoleServiceConstants.GLOBAL will be included.
   *
   * If group name is not specified, all roles including those assigned to
   * RoleServiceConstants.GLOBAL are returned.
   *
   * @param {string} userId
   * @param {string} group
   * @return {Promise<{error?: string; roles?: string[]}>}
   */
  getRolesForUser (input: { userId: string, group?: string }): Promise<{ error?: string, roles?: string[] }>

  getUsersInRoles (input: {
    roles: string | string[], group?: string, limit?: number,
    cursor?: string
  }): Promise<FindWithCursor<User>>

  removeUsersFromRoles (input: {
    userIds: string | string[], roles: string | string[],
    group?: string
  }): Promise<{ error?: string }>

  removeUsersFromAllRoles (input: { userIds: string | string[] }): Promise<{ error?: string }>

  /**
   * Returns true if user is assigned to one or more of the specified rule(s). This can be further
   * restricted based on group name.
   *
   * @param {{userId: string; roles: (string | string[]); group?: string}} input
   * @return {Promise<boolean>}
   */
  userIsInRoles (input: { userId: string, roles: string | string[], group?: string }): Promise<boolean>
}
