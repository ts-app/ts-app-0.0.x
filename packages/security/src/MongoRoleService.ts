import { RoleService, RoleServiceConstants } from './RoleService'
import { Role } from './Role'
import { ObjectId } from 'mongodb'
import { FindWithCursor, MongoService } from '@ts-app/mongo'
import { escapeRegex, User } from '@ts-app/common'
import { Resolver, ServiceInfo, userIdMatchParam } from '@ts-app/graphql'
import { DefaultRoles } from './DefaultRoles'

/**
 * Concept & implementation based on alanning/roles-npm
 */
export class MongoRoleService implements RoleService, ServiceInfo {
  private mongoService: MongoService

  constructor ({ mongoService }: { mongoService: MongoService }) {
    this.mongoService = mongoService
  }

  info (): { serviceName: string; } {
    return {
      serviceName: 'RoleService'
    }
  }

  @Resolver({ auth: [ DefaultRoles.Administrator ], type: 'mutation' })
  async addUsersToRoles ({ userIds, roles, group = RoleServiceConstants.GLOBAL }: {
    userIds: string | string[], roles: string | string[], group?: string
  }): Promise<{ error?: string }> {
    if (!userIds || !roles || !group) {
      return { error: 'Invalid arguments for addUsersToRoles()' }
    }

    const _userIds = typeof userIds === 'string' ? [ userIds ] : userIds
    const _roles = typeof roles === 'string' ? [ roles ] : roles
    let _group = group

    if (_group) {
      if (_group.startsWith('$')) {
        return { error: 'Group cannot start with \'$\'' }
      }
      _group = _group.replace(/\./g, '_')
    }

    const rolesWithGroup = _roles.map(role => ({ role, group: _group }))

    // --- create missing roles in roles collection
    const cursor = await this.mongoService.cursor<Role>('roles', { name: { $in: _roles } })
    const matchingRolesInDb = await cursor.toArray()
    const rolesNotInDb = _roles.filter(roleName => !matchingRolesInDb.find(roleInDb => roleInDb.name === roleName))
    for (const roleName of rolesNotInDb) {
      const createRole = await this.createRole({ name: roleName })
      if (!createRole) {
        return { error: `Error creating role [${roleName}]` }
      }
    }

    await cursor.close()

    // --- assign roles to users collection
    const Users = await this.mongoService.collection('users')
    const updateMany = await Users.updateMany(
      {
        _id: { $in: _userIds.map(id => new ObjectId(id)) }
      },
      {
        $addToSet: {
          roles: {
            $each: rolesWithGroup
          }
        }
      }
    )

    if (updateMany.result.ok === 1) {
      return {}
    } else {
      return { error: 'Error adding users to roles' }
    }
  }

  @Resolver({ auth: [ DefaultRoles.Administrator ], type: 'mutation' })
  async createRole ({ name }: { name: string }): Promise<{ error?: string, id: string }> {
    const id = await this.mongoService.create('roles', { name })
    return { id }
  }

  @Resolver({ auth: [ DefaultRoles.Administrator ], type: 'mutation' })
  async removeRole ({ name }: { name: string }): Promise<{ error?: string }> {
    const userWithRole = await this.mongoService.get('users',
      {
        'roles.role': name
      },
      { fields: { _id: 1 } }
    )

    if (userWithRole) {
      return { error: `Role [${name}] is in use` }
    } else {
      await this.mongoService.remove('roles', { name })
      return {}
    }
  }

  @Resolver({ auth: userIdMatchParam('userId') })
  async getGroupsForUser ({ role, userId }: { userId: string, role?: string }): Promise<{ error?: string, groups?: string[] }> {
    let filter

    if (role) {
      filter = {
        _id: new ObjectId(userId),
        'roles.role': role
      }
    } else {
      filter = { _id: new ObjectId(userId) }
    }

    const user = await this.mongoService.get<User>('users', filter, {
      fields: {
        'roles': 1
      }
    })

    if (!user) {
      if (role) {
        return { error: `Error getting groups for user [${userId}] with role [${role}]` }
      } else {
        return { error: `Error getting groups for user [${userId}]` }
      }
    } else if (role) {
      const roles = user.roles || []
      const groupNames = roles.filter(currentRole => (currentRole.role === role && currentRole.group !== RoleServiceConstants.GLOBAL))
        .map(currentRole => {
          return currentRole.group
        })

      return { groups: Array.from(new Set(groupNames)) }
    } else {
      const roles = user.roles || []
      const groupNames = roles.filter(currentRole => (currentRole.group !== RoleServiceConstants.GLOBAL))
        .map(currentRole => currentRole.group)

      return { groups: Array.from(new Set(groupNames)) }
    }
  }

  @Resolver({ auth: [ DefaultRoles.Administrator ] })
  findRolesWithCursor (input: { q?: string, limit?: number, cursor?: string } = {}): Promise<FindWithCursor<Role>> {
    const { q, limit, cursor } = input
    let filter = {}
    if (q && q.trim().length > 0) {
      filter = {
        $or: [
          { name: { $regex: `^${escapeRegex(q)}`, $options: 'i' } },
          { name: { $regex: `${escapeRegex(q)}$`, $options: 'i' } }
        ]
      }
    }

    return this.mongoService.findWithCursor<Role>('roles', filter, limit, cursor)
  }

  @Resolver({ auth: userIdMatchParam('userId'), name: 'userRoles' })
  async getRolesForUser ({ userId, group }: { userId: string, group?: string }): Promise<{ error?: string, roles?: string[] }> {
    let filter
    if (!group) {
      filter = { _id: new ObjectId(userId) }
    } else {
      filter = {
        _id: new ObjectId(userId),
        'roles.group': { $in: [ RoleServiceConstants.GLOBAL, group ] }
      }
    }

    const user = await this.mongoService.get<User>('users', filter, {
      fields: {
        'roles': 1
      }
    })

    if (!user) {
      return { error: `Error getting roles for user [${userId}]` }
    } else if (user.roles) {
      // user with roles
      const roles = user.roles
        .filter(currentRole => {
          // filter for global groups
          // if groups is specified, filter by current role's group
          return currentRole.group === RoleServiceConstants.GLOBAL || !group || currentRole.group === group
        })
        .map(currentRole => currentRole.role)

      return { roles: Array.from(new Set(roles)) }
    } else {
      // user without roles
      return { roles: [] }
    }
  }

  @Resolver({ auth: [ DefaultRoles.Administrator ] })
  async getUsersInRoles ({ group, roles, limit, cursor }: {
    roles: string | string[], group?: string, limit?: number, cursor?: string
  }): Promise<FindWithCursor<User>> {
    const _roles = typeof roles === 'string' ? [ roles ] : roles
    const _groups = group ? [ group, RoleServiceConstants.GLOBAL ] : [ RoleServiceConstants.GLOBAL ]

    let filter
    if (group) {
      filter = {
        roles: {
          $all: [
            { $elemMatch: { role: { $in: _roles }, group: { $in: _groups } } }
          ]
        }
      }
    } else {
      filter = {
        'roles.role': { $in: _roles }
      }
    }
    return this.mongoService.findWithCursor<User>('users', filter, limit, cursor)
  }

  @Resolver({ auth: [ DefaultRoles.Administrator ], type: 'mutation' })
  async removeUsersFromRoles ({ userIds, roles, group }: {
    userIds: string | string[], roles: string | string[], group?: string
  }): Promise<{ error?: string }> {
    const _userIds = typeof userIds === 'string' ? [ userIds ] : userIds
    const _roles = typeof roles === 'string' ? [ roles ] : roles
    const filter = {
      _id: { $in: _userIds.map(id => new ObjectId(id)) }
    }
    const update = {
      $pull: {
        roles: {
          role: { $in: _roles }
        }
      }
    }

    const collection = await this.mongoService.collection('users')
    const updateMany = await collection.updateMany(filter, update)

    if (updateMany.result.ok === 1) {
      return {}
    } else {
      return { error: `Error removing role(s) ${_roles} for user(s) ${_userIds}` }
    }

    // TODO: unused roles from collection('roles')
  }

  @Resolver({ auth: [ DefaultRoles.Administrator ], type: 'mutation' })
  async removeUsersFromAllRoles ({ userIds }: { userIds: string | string[] }): Promise<{ error?: string }> {
    const _userIds = typeof userIds === 'string' ? [ userIds ] : userIds
    const filter = {
      _id: { $in: _userIds.map(id => new ObjectId(id)) }
    }
    const update = {
      $unset: { roles: '' }
    }

    const collection = await this.mongoService.collection('users')
    const updateMany = await collection.updateMany(filter, update)

    if (updateMany.result.ok === 1) {
      return {}
    } else {
      return { error: `Error removing all roles for user(s) ${_userIds}` }
    }
  }

  @Resolver({ auth: userIdMatchParam('userId') })
  async userIsInRoles ({ userId, roles, group }: { userId: string, roles: string | string[], group?: string }): Promise<boolean> {
    try {
      const _role = typeof roles === 'string' ? [ roles ] : roles
      const _groups = group ? [ group, RoleServiceConstants.GLOBAL ] : [ RoleServiceConstants.GLOBAL ]

      const filter = {
        _id: new ObjectId(userId),
        roles: {
          $all: [
            { $elemMatch: { role: { $in: _role }, group: { $in: _groups } } }
          ]
        }
      }

      const findWithCursor = await this.mongoService.findWithCursor<User>('users', filter)
      return findWithCursor.docs.length > 0
    } catch (e) {
      // whenever error happens, return as user not in role
      console.error(e)
      return false
    }
  }
}
