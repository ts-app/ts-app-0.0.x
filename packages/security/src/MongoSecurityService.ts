import * as passport from 'passport'
import { StrategyOptions, Strategy, ExtractJwt } from 'passport-jwt'
import { Request, Response, NextFunction } from 'express'
import * as jwt from 'jsonwebtoken'
import * as bcryptjs from 'bcryptjs'
import { FindWithCursor, MongoService } from '@ts-app/mongo'
import { escapeRegex, Profile, User } from '@ts-app/common'
import { validateEmail, validatePassword } from './validators'
import { SecurityService } from './SecurityService'
import {
  Resolver,
  SecurityMiddleware,
  ServiceInfo,
  userIdMatchParam,
  environmentVariableIsSet
} from '@ts-app/graphql'
import { RoleService } from './RoleService'
import { DefaultRoles } from './DefaultRoles'

/**
 * SecurityService implementation that uses MongoDB as the data storage.
 *
 * This implementation also provides an Express middleware that configures JWT authentication.
 */
export class MongoSecurityService implements SecurityService, SecurityMiddleware, ServiceInfo {
  private mongoService: MongoService
  private roleService: RoleService
  private secretOrKey: string
  private accessTokenExpiresIn: string
  private refreshTokenExpiresIn: string

  constructor ({ mongoService, roleService, secretOrKey = 'taktau', accessTokenExpiresIn = '30m', refreshTokenExpiresIn = '30d' }: {
    mongoService: MongoService, roleService: RoleService
    secretOrKey?: string, accessTokenExpiresIn?: string, refreshTokenExpiresIn?: string
  }) {
    this.mongoService = mongoService
    this.roleService = roleService
    this.secretOrKey = secretOrKey
    this.accessTokenExpiresIn = accessTokenExpiresIn
    this.refreshTokenExpiresIn = refreshTokenExpiresIn

    this.configurePassport()
  }

  info (): { serviceName: string; } {
    return {
      serviceName: 'SecurityService'
    }
  }

  /**
   *
   * @param {string[]} roles List of roles authorized to access resource. Empty array allows any
   * authenticated request to access resource. Undefined/Null value allows authenticated requests
   * to access resource.
   * @return {(req: Request, res: Response, next: e.NextFunction) => any}
   */
  middleware (roles?: string[] | undefined) {
    const mustBeLoggedIn = !!roles
    const authorizedRoles = roles || []

    return (req: Request, res: Response, next: NextFunction) => {
      passport.authenticate('jwt', async (err: any, user: User | false, info: any) => {
        if (err) {
          console.error(err)
          res.status(500).send('Authentication error!')
        } else {
          if (user) {
            // --- request auth: step 2 of 3. assign user to request if authentication was successful
            req.user = user
            if (authorizedRoles && authorizedRoles.length > 0) {
              const isAuthorized = await this.roleService.userIsInRoles({
                userId: user.id,
                roles: authorizedRoles
              })
              if (isAuthorized) {
                // allow authenticated AND authorized access
                next()
              } else {
                // authenticated but not authorized (no access)
                res.status(401).send('Unauthorized access!')
              }
            } else {
              // allow any authenticated access
              next()
            }
          } else {
            // not authenticated
            if (mustBeLoggedIn) {
              res.status(401).send('Unauthenticated access!')
            } else {
              // allow unauthenticated access
              next()
            }
          }
        }
      })(req, res, next)
    }
  }

  configurePassport () {
    const loginWithUserId = this.loginWithUserId.bind(this)
    const jwtFromRequest = ExtractJwt.fromExtractors([
      ExtractJwt.fromAuthHeaderAsBearerToken(), ExtractJwt.fromUrlQueryParameter('accessToken')
    ])
    const strategyOptions: StrategyOptions = { jwtFromRequest, secretOrKey: this.secretOrKey }
    const strategy = new Strategy(strategyOptions, async (payload, next) => {
      try {
        // --- request auth: step 1 of 3. extract JWT token and perform user login if need
        const user = await loginWithUserId(payload.userId)
        if (!user.error) {
          // valid user login
          next(null, user.user)
        } else {
          // invalid user login
          next(null, false)
        }
      } catch (e) {
        console.error('Authentication error.', e)
        next(true)
      }
    })
    passport.use(strategy)
  }

  @Resolver({ type: 'mutation' })
  async signUp (input: { email: string, password: string }): Promise<{ error?: string, user?: User }> {
    const { email, password } = input

    if (!validateEmail(email)) {
      return { error: 'Invalid email' }
    }
    if (!validatePassword(password)) {
      return { error: 'Invalid password' }
    }

    const emailExist = await this.mongoService.get<User>('users', { 'emails.email': email })
    if (emailExist) {
      return { error: 'User already exist' }
    }

    const bcrypt = await bcryptjs.hash(password, 8)
    const userId = await this.mongoService.create<User>('users', {
      id: '',
      createdAt: new Date(),
      services: {
        password: {
          bcrypt
        }
      },
      emails: [
        { email, verified: true }
      ],
      profile: {
        email: email,
        name: email.substr(0, email.indexOf('@'))
      }
    })

    const user = await this.mongoService.get<User>('users', userId)
    if (user) {
      return { user }
    } else {
      return { error: `Error getting user with ID [${userId}]` }
    }
  }

  @Resolver({ type: 'mutation' })
  async loginWithEmailPassword (input: {
    email: string, password: string
  }): Promise<{ error?: string, userId?: string, accessToken?: string, refreshToken?: string }> {
    const { email, password } = input
    const filter = {
      emails: {
        email,
        verified: true
      }
    }
    const user = await this.mongoService.get<User>('users', filter)
    if (user) {
      const { services } = user
      const userBcrypt = services && services.password && services.password.bcrypt
      if (userBcrypt && await bcryptjs.compare(password, userBcrypt)) {
        // login successful, return tokens with user ID as payload
        const payload = {
          userId: user.id
        }
        return {
          userId: user.id,
          ...this.generateTokens(payload)
        }
      }
    }
    return { error: 'Invalid login attempt' }
  }

  @Resolver({ auth: [ DefaultRoles.Administrator ], type: 'query', paramNames: [ 'id' ] })
  async user (id: string): Promise<{ error?: string, user?: User }> {
    const user = await this.mongoService.get<User>('users', id)
    if (!user) {
      return { error: `Error getting user [${id}]` }
    } else {
      return { user }
    }
  }

  /**
   * Create test/seed users within the system.
   *
   * Resolver authorization prevents execution of this function unless environment variable 'seed' is set to '1'.
   */
  @Resolver({ auth: environmentVariableIsSet('seed'), type: 'mutation' })
  async seedUsers (input: { force?: boolean, userCount?: number } = {}): Promise<{ error?: string }> {
    const { force = false, userCount = 20 } = input
    if (!force) {
      // prevent seeding if users exist
      const Users = await this.mongoService.collection('users')
      const userCount = await Users.count({})

      if (userCount > 0) {
        return { error: 'Cannot seed database with users' }
      }
    } else {
      // remove seeded users
      await this.mongoService.remove('users', {
        $or: [
          { 'emails.email': { $regex: 'user[0-9]+@test\.local' } },
          { 'emails.email': 'admin@test.local' }
        ]
      })
    }

    let signUp = await this.signUp({ email: 'admin@test.local', password: 'testAdmin' })
    if (signUp.error) {
      return signUp
    }

    await this.roleService.addUsersToRoles({
      userIds: signUp!.user!.id,
      roles: DefaultRoles.Administrator
    })

    const userIds = []
    for (let c = 1; c <= userCount; c++) {
      let signUp = await this.signUp({ email: `user${c}@test.local`, password: 'testUser' })
      if (signUp.error) {
        return signUp
      }
      userIds.push(signUp!.user!.id)
    }
    await this.roleService.addUsersToRoles({ userIds, roles: DefaultRoles.User })

    return {}
  }

  @Resolver({ auth: [ DefaultRoles.Administrator ], type: 'query' })
  users (input: { q?: string, limit?: number, cursor?: string } = {}): Promise<FindWithCursor<User>> {
    const { q, limit = 10, cursor } = input
    let filter = {}
    if (q && q.trim().length > 0) {
      filter = {
        $or: [
          { 'profile.email': { $regex: `^${escapeRegex(q)}`, $options: 'i' } },
          { 'profile.email': { $regex: `${escapeRegex(q)}$`, $options: 'i' } },

          { 'profile.name': { $regex: `^${escapeRegex(q)}`, $options: 'i' } },
          { 'profile.name': { $regex: `${escapeRegex(q)}$`, $options: 'i' } }
        ]
      }
    }

    return this.mongoService.findWithCursor<User>('users', filter, limit, cursor)
  }

  @Resolver({ auth: [ DefaultRoles.Administrator ], type: 'mutation', paramNames: [ 'id' ] })
  async removeUser (id: String) {
    return this.mongoService.remove('users', id)
  }

  @Resolver({ auth: userIdMatchParam('id'), type: 'mutation' })
  async updateProfile (input: { id: string, profile: { [key: string]: any } }): Promise<{ error?: string }> {
    const { id, profile } = input
    const profileWithPrefix = Object.keys(profile).reduce((p: { [key: string]: any }, key: string) => {
      p[ `profile.${key}` ] = profile[ key ]
      return p
    }, {})

    await this.mongoService.update('users', id, {
      $set: {
        ...profileWithPrefix
      }
    })

    return {}
  }

  @Resolver({ auth: [ DefaultRoles.Administrator ], type: 'mutation' })
  async reset (): Promise<{ error?: string }> {
    const users = await this.mongoService.findWithCursor<User>('users', {
      'emails.email': 'admin@test.local'
    })

    if (users.docs.length === 0) {
      return { error: 'Pre-requisite to reset not met' }
    }

    return this.mongoService.dropCollection('users')
  }

  // --- stuff below... TODO: not done...

  @Resolver({ auth: userIdMatchParam('id'), type: 'mutation' })
  async setProfile<T extends Profile> (input: { id: string, profile: T }): Promise<void> {
    const { id, profile } = input
    return this.mongoService.update('users', id, { profile })
  }

  async getProfile<T extends Profile> (input: { id: string }): Promise<T> {
    const { id } = input
    const user = await this.mongoService.get<User>('users', id, {
      fields: { profile: 1 }
    })
    if (!user) {
      throw new Error(`Error getting profile for user [${id}]`)
    }

    return user.profile as T
  }

  async loginWithUserId (id: string): Promise<{ error?: string, user?: User }> {
    const user = await this.mongoService.get<User>('users', id)
    if (user) {
      return { user }
    }
    return { error: 'Invalid login attempt' }
  }

  verifyToken (token: string): any {
    return jwt.verify(token, this.secretOrKey)
  }

  private generateTokens (payload: any) {
    const accessToken = jwt.sign(payload, this.secretOrKey, {
      expiresIn: this.accessTokenExpiresIn
    })
    // TODO use var randtoken = require('rand-token')
    const refreshToken = jwt.sign(payload, this.secretOrKey, {
      expiresIn: this.refreshTokenExpiresIn
    })

    return { accessToken, refreshToken }
  }
}
