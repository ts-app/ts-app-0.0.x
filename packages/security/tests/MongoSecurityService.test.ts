import { MongoService } from '@ts-app/mongo'
import { MongoSecurityService } from '../src/MongoSecurityService'
import { SecurityService } from '../src/SecurityService'
import { Profile, User } from '@ts-app/common'
import { RoleService } from '../src/RoleService'
import { MongoRoleService } from '../src/MongoRoleService'

describe('MongoSecurityService', async () => {
  const localUrl = 'mongodb://localhost:27017'
  let mongoService: MongoService
  let roleService: RoleService
  let securityService: SecurityService

  beforeEach(async () => {
    mongoService = new MongoService(localUrl)
    roleService = new MongoRoleService({ mongoService })
    securityService = new MongoSecurityService({ mongoService, roleService })

    try {
      await mongoService.dropCollection('users')
    } catch {
      // ignore if cannot drop collection
    }
  })

  afterEach(async () => {
    const db = await mongoService.db()
    return db.close()
  })

  const createTestFixtures = async (): Promise<User> => {
    const signUp = await securityService.signUp({ email: 'test@test.com', password: 'abc123' })
    expect(signUp.error).toBeFalsy()
    expect(signUp.user!.id.length).toBe(24)
    return signUp.user!
  }

  test('user sign up & login with email/password (via bcrypt)', async () => {
    const user = await createTestFixtures()

    const loginWithEmailPassword = await securityService.loginWithEmailPassword({
      email: 'test@test.com',
      password: 'abc123'
    })
    expect(loginWithEmailPassword.error).toBeFalsy()

    const accessToken: any = (securityService as MongoSecurityService).verifyToken(loginWithEmailPassword.accessToken!)
    expect(accessToken.userId).toBe(user.id)

    const refreshToken: any = (securityService as MongoSecurityService).verifyToken(loginWithEmailPassword.refreshToken!)
    expect(refreshToken.userId).toBe(user.id)
  })

  test('user sign up & login with user ID (via JWT)', async () => {
    const user = await createTestFixtures()

    const loginWithUserId = await securityService.loginWithUserId(user.id)
    return expect(loginWithUserId.user).toMatchObject(user)
  })

  test('getProfile()', async () => {
    const user = await createTestFixtures()
    const sampleProfile: Profile = {
      name: 'aaa',
      email: 'aaa@bbb.com',
      avatarUrl: 'http://aaa'
    }

    let getProfile = await securityService.getProfile({ id: user.id })
    expect(getProfile).toMatchSnapshot()

    await securityService.setProfile({ id: user.id, profile: sampleProfile })
    getProfile = await securityService.getProfile({ id: user.id })
    expect(getProfile).toMatchObject(sampleProfile)

    const updatedProfile: Profile = { name: 'bbb', email: 'bbb@bbb.com' }
    await securityService.setProfile({ id: user.id, profile: updatedProfile })
    getProfile = await securityService.getProfile({ id: user.id })
    expect(getProfile).toMatchObject(updatedProfile)
  })

  test('setProfile() and updateProfile()', async () => {
    const user = await createTestFixtures()
    const sampleProfile: Profile = {
      name: 'aaa',
      email: 'aaa@bbb.com',
      avatarUrl: 'http://aaa'
    }

    await securityService.setProfile({ id: user.id, profile: sampleProfile })
    let getProfile = await securityService.getProfile({ id: user.id })
    expect(getProfile).toMatchObject(sampleProfile)

    // perform partial profile update
    await securityService.updateProfile({
      id: user.id,
      profile: {
        email: 'ccc@bbb.com'
      }
    })
    getProfile = await securityService.getProfile({ id: user.id })
    expect(getProfile).toMatchObject({
      ...sampleProfile,
      email: 'ccc@bbb.com'
    })
  })

  test('user()', async () => {
    const user = await createTestFixtures()
    const retrievedUser = await securityService.user(user.id)
    expect(retrievedUser.user!).toMatchObject(user)
  })

  test('user() with invalid user ID', async () => {
    const retrievedUser = await securityService.user('012345678912')
    expect(retrievedUser).toMatchSnapshot()
  })

  test('seedUsers()', async () => {
    const seedUsers = await securityService.seedUsers()
    expect(seedUsers.error).toBeFalsy()

    // users created
    const users = await securityService.users({ limit: 100 })
    expect(users.docs.length).toBe(21)

    // can login as admin, can get access token
    let login = await securityService.loginWithEmailPassword({
      email: 'admin@test.local',
      password: 'testAdmin'
    })
    expect(login.error).toBeFalsy()
    expect(login.accessToken!.length).toBeGreaterThan(0)

    // can login as test user
    login = await securityService.loginWithEmailPassword({
      email: 'user1@test.local',
      password: 'testUser'
    })
    expect(login.error).toBeFalsy()
    expect(login.accessToken!.length).toBeGreaterThan(0)
  })

  test('seedUsers() cannot seed twice', async () => {
    const seedUsers = await securityService.seedUsers()
    expect(seedUsers.error).toBeFalsy()

    const reseedUsers = await securityService.seedUsers()
    expect(reseedUsers.error).toBeTruthy()
  })

  test('users()', async () => {
    const seedUsers = await securityService.seedUsers()
    expect(seedUsers.error).toBeFalsy()

    const admin = await securityService.users({ q: 'admi' })
    expect(admin.docs.length).toBe(1)

    const users = await securityService.users({ q: 'use' })
    expect(users.docs.length).toBe(10)
    expect(users.cursor).toBeTruthy()

    const users11to20 = await securityService.users({ q: 'use', cursor: users.cursor })
    expect(users11to20.docs.length).toBe(10)
    expect(users11to20.cursor).toBeTruthy()

    const users21to30 = await securityService.users({ q: 'use', cursor: users11to20.cursor })
    expect(users21to30.cursor).toBeFalsy()
    expect(users21to30.docs.length).toBe(0)
  })

  test('deleteUser()', async () => {
    const user = await createTestFixtures()
    expect(await securityService.removeUser(user.id)).toBe(1)
    // delete again...
    expect(await securityService.removeUser(user.id)).toBe(0)
  })

  test('seedUsers(true) will force delete', async () => {
    const seedUsers = await securityService.seedUsers()
    expect(seedUsers.error).toBeFalsy()

    const reseedUsers = await securityService.seedUsers({ force: true })
    expect(reseedUsers.error).toBeFalsy()
  })

  test('reset()', async () => {
    // create test user
    await createTestFixtures()

    // cannot reset if admin@test.local does not exist
    let reset = await securityService.reset()
    expect(reset.error).toBe('Pre-requisite to reset not met')

    // seed test user (admin@test.local)
    const seed = await securityService.seedUsers({ force: true })
    expect(seed.error).toBeFalsy()

    // reset database (a system with admin@test.local is seen as a test system, so allow reset())
    reset = await securityService.reset()
    expect(reset.error).toBeFalsy()
  })

  test('seedUsers() - create roles', async () => {
    const seedUsers = await securityService.seedUsers()
    expect(seedUsers.error).toBeFalsy()

    const adminLogin = await securityService.loginWithEmailPassword({
      email: 'admin@test.local',
      password: 'testAdmin'
    })
    let { userId } = (securityService as MongoSecurityService).verifyToken(adminLogin.accessToken!)
    let roles = await roleService.getRolesForUser({ userId })
    expect(roles).toMatchObject({ roles: [ 'Administrator' ] })

    let userLogin = await securityService.loginWithEmailPassword({
      email: 'user1@test.local',
      password: 'testUser'
    })
    userId = (securityService as MongoSecurityService).verifyToken(userLogin.accessToken!).userId
    roles = await roleService.getRolesForUser({ userId })
    expect(roles).toMatchObject({ roles: [ 'User' ] })
  })
})
