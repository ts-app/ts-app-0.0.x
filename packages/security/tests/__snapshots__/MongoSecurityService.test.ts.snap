// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MongoSecurityService getProfile() 1`] = `
Object {
  "email": "test@test.com",
  "name": "test",
}
`;

exports[`MongoSecurityService user() with invalid user ID 1`] = `
Object {
  "error": "Error getting user [012345678912]",
}
`;
